# HG changeset patch
# User Rik <rik@octave.org>
# Date 1445960496 25200
#      Tue Oct 27 08:41:36 2015 -0700
# Node ID 0140032737375fb8a2396667af36754f9b97b432
# Parent  63f975ff1f7c066ad8565acb279ce0c9bc17bfa5
isdiag.m: Add shortcut code for inputs which are diagonal matrices already.

* isdiag.m: Check if the typeinfo indicates the input is a diagonal matrix and
return true.  Rework BIST tests.

diff --git a/scripts/linear-algebra/isdiag.m b/scripts/linear-algebra/isdiag.m
--- a/scripts/linear-algebra/isdiag.m
+++ b/scripts/linear-algebra/isdiag.m
@@ -25,32 +25,35 @@
 ## Author: Massimiliano Fasi
 
 function retval = isdiag (A)
 
   if (nargin != 1)
     print_usage ();
   endif
 
-  retval = (isnumeric (A) || islogical (A)) && ndims (A) == 2;
-  if (retval)
+  if (strcmp (typeinfo (A), "diagonal matrix"))
+    retval = true;
+  elseif ((isnumeric (A) || islogical (A)) && ndims (A) == 2)
     [i, j] = find (A);
     retval = all (i == j);
+  else
+    retval = false;
   endif
 
 endfunction
 
 
-%!assert (! isdiag ("string"))
+%!assert (isdiag ("string"), false)
+%!assert (isdiag (zeros (2,2,2)), false)
 %!assert (isdiag ([]))
-
 %!assert (isdiag (1))
-%!assert (! isdiag ([1, 1]))
-%!assert (! isdiag ([1; 1]))
+%!assert (isdiag ([1, 1]), false)
+%!assert (isdiag ([1; 1]), false)
 %!assert (isdiag (eye (10)))
-%!assert (issymmetric ([i, 0; 0, 1 + i]))
-%!assert (isdiag (speye (1000000)))
 %!assert (isdiag (logical (eye (10))))
+%!assert (isdiag (speye (1e6)))
+%!assert (isdiag (diag (1:10)))
 
 ## Test input validation
 %!error isdiag ()
 %!error isdiag (1,2)
 
